[Temporal] Implement addition and subtraction for PlainYearMonth

catamorphism opened the pull request:
#### 70f276559133cbe9af745868ab2794d138d47849
<pre>
[Temporal] Implement addition and subtraction for PlainYearMonth
<a href="https://bugs.webkit.org/show_bug.cgi?id=304093">https://bugs.webkit.org/show_bug.cgi?id=304093</a>

Reviewed by Sosuke Suzuki.

Implement `add` and `subtract` methods.

Co-authored-by: SUZUKI Sosuke &lt;aosukeke@gmail.com&gt;

* JSTests/stress/temporal-plainyearmonth.js:
(shouldBe):
* JSTests/test262/config.yaml:
* Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp:
(JSC::TemporalPlainYearMonth::addDurationToYearMonth):
* Source/JavaScriptCore/runtime/TemporalPlainYearMonth.h:
* Source/JavaScriptCore/runtime/TemporalPlainYearMonthPrototype.cpp:
(JSC::temporalPlainYearMonthPrototypeAddOrSubtract):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/304464@main">https://commits.webkit.org/304464@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0e7ece2a09107c88ff783976615512ca419eeb0c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/135507 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7885 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46796 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143202 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/87190 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/8522 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7732 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103551 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/70900 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/138453 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6117 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121460 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84420 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5894 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3505 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/3811 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/127500 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115100 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39646 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145952 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/133991 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/7570 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40217 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111915 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7609 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6352 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112287 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28516 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5751 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117762 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/61480 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7622 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35878 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/166830 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7369 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71168 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43568 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7588 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/7470 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sosukesuzuki commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp:
```diff
@@ -235,4 +235,41 @@ String TemporalPlainYearMonth::monthCode() const
 
+// https://tc39.es/proposal-temporal/#sec-temporal-adddurationtoyearmonth
+ISO8601::PlainYearMonth TemporalPlainYearMonth::addDurationToYearMonth(JSGlobalObject* globalObject,
+    AddOrSubtract op, ISO8601::PlainYearMonth yearMonth, ISO8601::Duration duration, TemporalOverflow overflow)
```
I think we can define `AddOrSubtract op` as a template type parameter like [`trimString`](https://github.com/WebKit/WebKit/blob/4c8d5ef6787e2c149355f472271fc7f77ca9d670/Source/JavaScriptCore/runtime/StringPrototype.cpp#L1178-L1179)

catamorphism replied:
Done

sosukesuzuki commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp:
```diff
@@ -235,4 +235,41 @@ String TemporalPlainYearMonth::monthCode() const
+    auto sign = TemporalDuration::sign(duration);
+    auto year = yearMonth.year();
+    auto month = yearMonth.month();
+    auto day = 1;
```
Can be `constexpr` ?

catamorphism replied:
Done

sosukesuzuki commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonthPrototype.cpp:
```diff
@@ -106,6 +110,43 @@ void TemporalPlainYearMonthPrototype::finishCreation(VM& vm, JSGlobalObject*)
 }
 
+static JSC::EncodedJSValue
+temporalPlainYearMonthPrototypeAddOrSubtract(JSGlobalObject* globalObject, CallFrame* callFrame, AddOrSubtract op)
```
style nits: According to the existing code, we shouldn't put line terminator here.

catamorphism replied:
Fixed

sosukesuzuki commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonth.h:
```diff
@@ -47,7 +47,10 @@ class TemporalPlainYearMonth final : public JSNonFinalObject {
 
-    static TemporalPlainYearMonth* from(JSGlobalObject*, JSValue item, JSValue optionsValue);
+    static ISO8601::PlainYearMonth addDurationToYearMonth(
+        JSGlobalObject*, AddOrSubtract, ISO8601::PlainYearMonth, ISO8601::Duration, TemporalOverflow);
```
style nits: According to the existing code, we shouldn't put line terminator here.

catamorphism replied:
Fixed

sosukesuzuki commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp:
```diff
@@ -235,4 +235,41 @@ String TemporalPlainYearMonth::monthCode() const
 
+// https://tc39.es/proposal-temporal/#sec-temporal-adddurationtoyearmonth
+ISO8601::PlainYearMonth TemporalPlainYearMonth::addDurationToYearMonth(JSGlobalObject* globalObject,
+    AddOrSubtract op, ISO8601::PlainYearMonth yearMonth, ISO8601::Duration duration, TemporalOverflow overflow)
```
style nits: According to the existing code, we shouldn't put line terminator here.

catamorphism replied:
Fixed

sosukesuzuki commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonthPrototype.cpp:
```diff
@@ -106,6 +110,43 @@ void TemporalPlainYearMonthPrototype::finishCreation(VM& vm, JSGlobalObject*)
+
+    auto* yearMonth = jsDynamicCast<TemporalPlainYearMonth*>(callFrame->thisValue());
+    if (!yearMonth) [[unlikely]]
+        return throwVMTypeError(globalObject, scope, "Temporal.PlainYearMonth.prototype.add called on value that's not a PlainYearMonth"_s);
```
This error message is incorrect for `subtract`

catamorphism replied:
Fixed

sosukesuzuki commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonthPrototype.cpp:
```diff
@@ -106,6 +110,43 @@ void TemporalPlainYearMonthPrototype::finishCreation(VM& vm, JSGlobalObject*)
+
+    ISO8601::PlainYearMonth result =
+        TemporalPlainYearMonth::addDurationToYearMonth(
+            globalObject, op, yearMonth->plainYearMonth(), duration, overflow);
```
style nits: According to the existing code, we shouldn't put line terminator here.

catamorphism replied:
Fixed

sosukesuzuki commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp:
```diff
@@ -235,4 +235,41 @@ String TemporalPlainYearMonth::monthCode() const
+    VM& vm = globalObject->vm();
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    if (op == AddOrSubtract::Subtract)
```
```suggestion
    if constexpr (op == AddOrSubtract::Subtract)
```

sosukesuzuki commented on Source/JavaScriptCore/runtime/TemporalPlainYearMonthPrototype.cpp:
```diff
@@ -106,6 +110,43 @@ void TemporalPlainYearMonthPrototype::finishCreation(VM& vm, JSGlobalObject*)
+
+    auto* yearMonth = jsDynamicCast<TemporalPlainYearMonth*>(callFrame->thisValue());
+    if (!yearMonth) [[unlikely]] {
+        String message = op == AddOrSubtract::Add ? "Temporal.PlainYearMonth.prototype.add called on value that's not a PlainYearMonth"_s : "Temporal.PlainYearMonth.prototype.subtract called on value that's not a PlainYearMonth"_s;
```
Can you use `if constexpr` here, instead of ternary

sosukesuzuki replied:
I prefer
```cpp
if constepxr (op == AddOrSubtract::Add)
  return throwVMTypeError(globalObject, scope, "..."_s);
else
  return throwVMTypeError(globalObject, scope, "..."_s);
```

sosukesuzuki reviewed (approved):
r=me

catamorphism committed:
[Temporal] Implement addition and subtraction for PlainYearMonth
https://bugs.webkit.org/show_bug.cgi?id=304093

Reviewed by Sosuke Suzuki.

Implement `add` and `subtract` methods.

Co-authored-by: SUZUKI Sosuke <aosukeke@gmail.com>

* JSTests/stress/temporal-plainyearmonth.js:
(shouldBe):
* JSTests/test262/config.yaml:
* Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp:
(JSC::TemporalPlainYearMonth::addDurationToYearMonth):
* Source/JavaScriptCore/runtime/TemporalPlainYearMonth.h:
* Source/JavaScriptCore/runtime/TemporalPlainYearMonthPrototype.cpp:
(JSC::temporalPlainYearMonthPrototypeAddOrSubtract):
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/304464@main
